AffineTranform variables should be able to be constexpr

litherum opened the pull request:
#### 9aa7ca7e355a7c79be8fb60dc61f6ebea652c605
<pre>
AffineTranform variables should be able to be constexpr
<a href="https://bugs.webkit.org/show_bug.cgi?id=242323">https://bugs.webkit.org/show_bug.cgi?id=242323</a>

Reviewed by Cameron McCormack.

If we make the AffineTransform constructor constexpr by moving it into the header,
it will then be possible to have AffineTransform variables be constexpr.

The style checker fails because it says the parameter names in
AffineTransform(double a, double b, double c, double d, double e, double f)
are redundant. However, these names are terms-of-art for the values in an affine
transformation matrix.

* Source/WebCore/platform/graphics/SourceBrush.cpp:
(WebCore::SourceBrush::gradientSpaceTransform const):
* Source/WebCore/platform/graphics/coretext/FontCascadeCoreText.cpp:
(WebCore::rotateLeftTransform):
* Source/WebCore/platform/graphics/transforms/AffineTransform.cpp:
(WebCore::AffineTransform::AffineTransform): Deleted.
* Source/WebCore/platform/graphics/transforms/AffineTransform.h:
(WebCore::AffineTransform::AffineTransform):
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::localTransform const):
(WebCore::RenderObject::localToParentTransform const):
* Source/WebCore/rendering/svg/SVGRenderSupport.h:
* Source/WebCore/svg/properties/SVGAnimationAdditiveListFunctionImpl.h:
(WebCore::SVGAnimationTransformListFunction::animate):
* Tools/TestWebKitAPI/Tests/WebCore/AffineTransform.cpp:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/252127@main">https://commits.webkit.org/252127@main</a>
</pre>

heycam commented on Source/WebCore/rendering/RenderObject.cpp:
```diff
@@ -1947,13 +1947,13 @@ FloatRect RenderObject::repaintRectInLocalCoordinates() const
 {
-    static const AffineTransform identity;
+    static constexpr AffineTransform identity;
     return identity;
```
This function does not return a reference, so you could write `return AffineTransform();` instead.

heycam commented on Source/WebCore/platform/graphics/SourceBrush.cpp:
```diff
@@ -36,14 +36,14 @@ SourceBrush::SourceBrush(const Color& color, std::optional<Brush>&& brush)
 const AffineTransform& SourceBrush::gradientSpaceTransform() const
 {
-    static NeverDestroyed<AffineTransform> identity;
+    static constexpr AffineTransform identity;
```
I wouldn't be surprised if the compiler de-duplicated this and the one in `RenderObject::localToParentTransform`. But if you want to avoid having this line in both functions, you could add that

```
static constexpr AffineTransform identity;
```

as a public member variable of `AffineTransform`, then do `return AffineTransform::identity;` here and in `RenderObject::localToParentTransform`. (In C++17, a static constexpr data member is implicitly an "inline variable", which means you don't need to write the variable definition in the cpp file, and the compiler will merge all of the instances of the variable across translation units.)

litherum replied:
That's a good idea, but I can't put it inside `AffineTransform` as a member because of https://stackoverflow.com/questions/36863503/struct-is-non-literal-type/36863691#36863691. I'll put it in `WebCore` next to the `AffineTransform` class.

heycam replied:
That's annoying. I guess you'll need to mark them explicitly as `inline` since they'll no longer be static member variables.

heycam reviewed (approved):
r=me assuming the MSVC special case removal works out.

litherum committed:
AffineTranform variables should be able to be constexpr
https://bugs.webkit.org/show_bug.cgi?id=242323

Reviewed by Cameron McCormack.

If we make the AffineTransform constructor constexpr by moving it into the header,
it will then be possible to have AffineTransform variables be constexpr.

The style checker fails because it says the parameter names in
AffineTransform(double a, double b, double c, double d, double e, double f)
are redundant. However, these names are terms-of-art for the values in an affine
transformation matrix.

* Source/WebCore/platform/graphics/SourceBrush.cpp:
(WebCore::SourceBrush::gradientSpaceTransform const):
* Source/WebCore/platform/graphics/coretext/FontCascadeCoreText.cpp:
(WebCore::rotateLeftTransform):
* Source/WebCore/platform/graphics/transforms/AffineTransform.cpp:
(WebCore::AffineTransform::AffineTransform): Deleted.
* Source/WebCore/platform/graphics/transforms/AffineTransform.h:
(WebCore::AffineTransform::AffineTransform):
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::localTransform const):
(WebCore::RenderObject::localToParentTransform const):
* Source/WebCore/rendering/svg/SVGRenderSupport.h:
* Source/WebCore/svg/properties/SVGAnimationAdditiveListFunctionImpl.h:
(WebCore::SVGAnimationTransformListFunction::animate):
* Tools/TestWebKitAPI/Tests/WebCore/AffineTransform.cpp:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/252127@main
